File: 2018/nim/day08.nim
import math, sequtils, strutils

type NodeArray = object
    vec: seq[int]
    value: int
    sum_meta: int

proc eval_node(vec_arg: seq[int]): NodeArray =
    let (num_children, len_metadata) = (vec_arg[0], vec_arg[1])
    var vec = vec_arg[2..<len(vec_arg)]
    var (children, sum_meta) = (newSeq[NodeArray](), 0)
    for _ in 1..num_children:
        let child = eval_node(vec)
        vec = child.vec
        sum_meta += child.sum_meta
        children.add(child)
    let metadata = vec[0..<len_metadata]
    sum_meta += sum(metadata)
    var value = 0
    if num_children > 0:
        for idx in metadata:
            if 0 < idx and idx <= len(children):
                value += children[idx - 1].value
    else:
        value = sum(metadata)
    NodeArray(vec: vec[len_metadata..<len(vec)],
              value: value, sum_meta: sum_meta)

let file = readFile("data/input-08.txt").strip().splitLines()
let vals = file[0].split(" ").map(parseInt)
let root = eval_node(vals)

let res1 = root.sum_meta
echo "Part 1: ", res1
assert(res1 == 41849)

let res2 = root.value
echo "Part 2: ", res2
assert(res2 == 32487)

File: 2018/src/day08.rs
use std::fs;

struct NodeSummary<'a> {
    vec: &'a [u16],
    value: u16,
    sum_meta: u16,
}

fn eval_node(vec: &[u16]) -> NodeSummary<'_> {
    let (num_children, len_metadata) = (vec[0], vec[1]);
    let (mut children, mut sum_meta) = (Vec::new(), 0);
    let mut vec = vec.split_at(2).1;
    for _ in 0..num_children {
        let child = eval_node(vec);
        vec = child.vec;
        sum_meta += child.sum_meta;
        children.push(child);
    }
    let (metadata, vec) = vec.split_at(len_metadata as usize);
    let value = match num_children {
        0 => metadata.iter().cloned().sum(),
        _ => metadata.iter().fold(0, |acc, &z| {
            acc + match z <= num_children {
                true => children[z as usize - 1].value,
                false => 0,
            }
        }),
    };
    sum_meta += metadata.iter().cloned().sum::<u16>();
    NodeSummary {
        vec,
        value,
        sum_meta,
    }
}

pub fn day08() {
    let input = "data/input-08.txt";
    let license: Vec<_> = fs::read_to_string(input)
        .unwrap()
        .replace('\n', "")
        .split(' ')
        .map(|z| z.parse::<u16>().unwrap())
        .collect();
    let root = eval_node(&license);

    let res = root.sum_meta;
    println!("Part 1: {res}");
    assert_eq!(res, 41849);

    let res = root.value;
    println!("Part 2: {res}");
    assert_eq!(res, 32487);
}
